book/src/qobject/cpp_object.md

```diff
@@ -5,36 +5,106 @@ SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
+The instance of the Rust struct is constructed using its [`Default`](https://doc.rust-lang.org/std/default/macro.Default.html) implementation in the C++ constructor.
+Therefore implementing `Default` on the Rust struct is currently mandatory.
 
-```rust,ignore,noplayground
-{{#include ../../../examples/qml_features/rust/src/rust_obj_invokables.rs:book_cpp_obj}}
+The C++ object will defer any property state to the Rust struct, and is therefore only a thin wrapper.
+This means that any `Q_PROPERTY`s data is stored in the Rust struct and that the getters and setters will modify the Rust struct internally.
+
+Furthermore, CXX-Qt generates functions on this struct that help you interact with Qt functionality.
+This includes access to the [`CxxQtThread` struct](./cxxqtthread.md), as well as a function to emit signals.
+
+## Referencing another QObject
+
+The `qobject::T` type is defined both within the CXX-Qt bridge, as well as the surrounding module.
+The simplest way is to reference it from the surrounding module.
+
+Example:
+``` rust,ignore,noplayground
+// In file qt_types.rs
+#[cxx_qt::bridge]
+mod ffi {
+  #[cxx_qt::qobject]
+  #[derive(Default)]
+  pub struct MyObject {}
+}
 ```
 
-If there is a [`Signals` enum](./signals_enum.md) then you can call `emit_queued(&mut self, Signals)` on the `CppObj` to emit a signal.
+``` rust,ignore,noplayground
+// In another module
+fn my_function(parameter: &crate::qt_types::qobject::MyObject) { /* ... */ }
+```
+
+Unfortunately, nested QObjects aren't yet supported by CXX-Qt.
+Progress can be tracked in [#299](https://github.com/KDAB/cxx-qt/issues/299).
+
+## `impl qobject::T`
+As mentioned before, the C++ QObject is exposed to Rust as an opaque CXX type.
+A side-effect of this is that you can actually extend this type with your own functionality by using normal `impl` blocks.
+Because `qobject::T` is an opaque C++ type, the same rules as with normal CXX opaque types apply.
+Most importantly this means that the type may never be accessed by value or by mutable self reference directly.
+Rather, all methods must use either `&self` or `self: Pin<&mut Self>` as the self types.
 
-Note that signals are queued to avoid a deadlock if the `Q_EMIT` is `Qt::DirectConnection` connected to a Rust invokable on the same QObject that has caused the `Q_EMIT`, as this would then try to lock the internal Rust object which is already locked.
+For more information about pinning, see: https://doc.rust-lang.org/std/pin/
 
-```rust,ignore,noplayground
-{{#include ../../../examples/qml_features/rust/src/signals.rs:book_rust_obj_impl}}
+In addition to methods that extend the Rust interface of a `qobject::T`, you may also mark methods within an `impl qobject::T` block with the `#[qinvokable]` attribute.
+These methods will then be exposed to the C++ QObject and can then be called by C++ or QML.
+See the [QObject page](./qobject_struct.md#invokables) for more details.
```

Comment:
Are non-invokable methods not available to be called from C++ code?

Reply:
No, currently not.
We've considered adding this feature at some point, but it's unclear what the use case of "callable" from C++, but not QML is.
Possibly useful if the type isn't QML compatible :thinking:

Reply:
https://github.com/KDAB/cxx-qt/issues/298 / https://github.com/KDAB/cxx-qt/discussions/279 

It's also potentially from an API point of view that wants to only expose certain methods to QML and certain methods to C++.